mne/beamformer/_dics.py

```diff
@@ -1,27 +1,581 @@
+                # Compute the norm for each set of 3 dipoles
+                W = W.reshape(-1, 3, W.shape[1])
+                norm = np.sqrt(np.sum(W ** 2, axis=(1, 2)))
+                W /= norm[:, np.newaxis, np.newaxis]
+                W = W.reshape(-1, W.shape[2])
+            else:
+                # Compute the norm for each dipole
+                norm = np.sqrt(np.sum(W ** 2, axis=1))
+                W /= norm[:, np.newaxis]
+
+        Ws.append(W)
+
+    Ws = np.array(Ws)
+
+    subject = _subject_from_forward(forward)
+    filters = dict(weights=Ws, csd=csd, ch_names=ch_names, proj=proj,
+                   vertices=vertices, subject=subject,
+                   pick_ori=pick_ori, inversion=inversion,
+                   weight_norm=weight_norm,
+                   normalize_fwd=normalize_fwd, n_orient=n_orient
+                   if pick_ori is None else 1)
+
+    return filters
+
+
+def _apply_dics(data, filters, info, tmin):
+    """Apply DICS spatial filter to data for source reconstruction."""
+    if isinstance(data, np.ndarray) and data.ndim == 2:
+        data = [data]
+        one_epoch = True
+    else:
+        one_epoch = False
+
+    Ws = filters['weights']
+    one_freq = len(Ws) == 1
+
+    subject = filters['subject']
+    for i, M in enumerate(data):
+        if not one_epoch:
+            logger.info("Processing epoch : %d" % (i + 1))
+
+        # Apply SSPs
+        if info['projs']:
+            _check_proj_match(info, filters)
+            M = np.dot(filters['proj'], M)
+
+        stcs = []
+        for W in Ws:
+            # project to source space using beamformer weights
+            sol = np.dot(W, M)
+
+            if filters['n_orient'] > 1:
+                logger.info('combining the current components...')
+                sol = combine_xyz(sol)
+
+            tstep = 1.0 / info['sfreq']
+
+            stcs.append(_make_stc(sol, vertices=filters['vertices'], tmin=tmin,
+                                  tstep=tstep, subject=subject))
+        if one_freq:
+            yield stcs[0]
+        else:
+            yield stcs
+
+    logger.info('[done]')
+
+
+@verbose
+def apply_dics(evoked, filters, verbose=None):
+    """Apply Dynamic Imaging of Coherent Sources (DICS) beamformer weights.
+
+    Apply Dynamic Imaging of Coherent Sources (DICS) beamformer weights
+    on evoked data.
+
+    .. warning:: The result of this function is meant as an intermediate step
+                 for further processing (such as computing connectivity). If
+                 you are interested in estimating source time courses, use an
+                 LCMV beamformer (:func:`make_lcmv`, :func:`apply_lcmv`)
+                 instead. If you are interested in estimating spectral power at
+                 the source level, use :func:`dics_source_power`.
+    .. warning:: This implementation has not been heavily tested so please
+                 report any issues or suggestions.
```

Comment:
the `apply_dics` and `apply_dics_epochs` apply the DICS filters directly to data. This can be valid in some cases, but we never do this in our lab. These functions remain there for backwards compatibility.